lark/parsers/lalr_parser.py

```diff
@@ -120,12 +120,19 @@ def feed_token(self, token, is_end=False):
         end_state = self.parse_conf.end_state
         callbacks = self.parse_conf.callbacks
 
+        # Make a copy of state_stack
+        original_state_stack = list(state_stack)
```

Comment:
I don't like unconditionally copying the stacks, but I am not sure if there is a better solution. Maybe at least delay it till the first `Reduce` action?

Reply:
Changed. Is this better?

Reply:
It at least reduces the performance hit a bit. But I am still not sure if it is worth it, since it is an unconditional performance hit only for error messages. Can try to run these benchmarks: https://github.com/goodmami/python-parsing-benchmarks for lark?

Reply:
I'm not sure how to run the benchmarks with my local copy, so what I did is to override lalr_parser.py in the venv directory (+ making sure this file is indeed used).

Before the change:

```
-------------------------------------- benchmark 'arithmetic': 1 tests ---------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     1.8369  1.9139  1.8743  0.0347  1.8891  0.0600       3;0  0.5335       5           1
--------------------------------------------------------------------------------------------------------------

------------------------------------- benchmark 'arithmetic-compile': 1 tests --------------------------------------
Name (time in ms)              Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     9.4613  20.4742  11.0787  2.5286  10.1461  0.4207     10;18  90.2630      98           1
--------------------------------------------------------------------------------------------------------------------

----------------------------------------------- benchmark 'ini': 1 tests -----------------------------------------------
Name (time in ms)              Min       Max      Mean   StdDev    Median      IQR  Outliers     OPS  Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     311.2880  344.4021  332.7991  15.1623  341.6660  24.8145       1;0  3.0048       5           1
------------------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'ini-compile': 1 tests -----------------------------------------
Name (time in ms)              Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     9.6849  19.4452  11.0329  2.1035  10.3357  0.2728     10;19  90.6381      96           1
--------------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json': 1 tests ------------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     3.5411  3.6372  3.5782  0.0378  3.5741  0.0521       1;0  0.2795       5           1
--------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     15.0438  28.1825  17.5996  3.1443  16.3627  0.4059     10;12  56.8194      62           1
---------------------------------------------------------------------------------------------------------------------

```

After the change:

```
-------------------------------------- benchmark 'arithmetic': 1 tests ---------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     2.0184  2.0845  2.0559  0.0293  2.0669  0.0513       1;0  0.4864       5           1
--------------------------------------------------------------------------------------------------------------

------------------------------------- benchmark 'arithmetic-compile': 1 tests --------------------------------------
Name (time in ms)              Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     9.6690  20.8623  11.5581  2.2187  10.7658  0.3369     12;21  86.5194      93           1
--------------------------------------------------------------------------------------------------------------------

----------------------------------------------- benchmark 'ini': 1 tests -----------------------------------------------
Name (time in ms)              Min       Max      Mean   StdDev    Median      IQR  Outliers     OPS  Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     339.6006  376.2953  357.4534  16.1469  353.9490  29.1214       2;0  2.7976       5           1
------------------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'ini-compile': 1 tests -----------------------------------------
Name (time in ms)              Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     9.7191  19.5331  11.6324  1.9314  11.0307  0.2571      9;17  85.9665      91           1
--------------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json': 1 tests ------------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     3.9242  3.9995  3.9707  0.0324  3.9854  0.0521       1;0  0.2518       5           1
--------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     16.5646  26.4917  18.4586  2.2015  17.6902  0.4268      7;12  54.1753      58           1
---------------------------------------------------------------------------------------------------------------------

```

Maybe we can do it under a flag (e.g., `accurate_exceptions`)?
This bug is currently preventing us from using the LALR mode, because we need good error messages for the users.

Reply:
Maybe it can be implemented via a flag, and then an automatic reparsing happens when an error occurs, **and** the users requests to use `.interactive_parser` (this also happens when `.accepts` is accessed to print the error message)?  But I don't think a 10% performance hit for successful execution is worth it. (Thank you for running the benchmark. I think the answer would have been `pip install -e .` to install the local package, but you figured it out anyway.)

Reply:
I added a flag that's by default False. If someone (like me) cares about the errors enough to lose some performance they can turn it on.

I redid the benchmarks since they were noisy. All the numbers now are higher because I fixed my CPU freq to 2GHz.
By the way, if you really care about performance I think you should remove the asserts in the critical path. See the last benchmark.

Before the change:

```
-------------------------------------- benchmark 'arithmetic': 1 tests ---------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     3.4774  3.5526  3.5111  0.0270  3.5081  0.0255       2;0  0.2848       5           1
--------------------------------------------------------------------------------------------------------------

-------------------------------------- benchmark 'arithmetic-compile': 1 tests --------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     18.8610  33.4660  19.9224  2.7566  19.0908  0.6966       2;3  50.1948      38           1
---------------------------------------------------------------------------------------------------------------------

----------------------------------------------- benchmark 'ini': 1 tests -----------------------------------------------
Name (time in ms)              Min       Max      Mean   StdDev    Median      IQR  Outliers     OPS  Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     577.3088  611.5630  593.7888  13.3596  595.8848  20.0004       2;0  1.6841       5           1
------------------------------------------------------------------------------------------------------------------------

------------------------------------------ benchmark 'ini-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     19.1115  32.0133  20.3093  2.4676  19.3569  1.3748       4;4  49.2386      51           1
---------------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json': 1 tests ------------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     6.4420  6.4658  6.4540  0.0092  6.4551  0.0138       2;0  0.1549       5           1
--------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     29.5193  40.9094  31.5422  2.9765  29.9385  2.7600       4;3  31.7035      32           1
---------------------------------------------------------------------------------------------------------------------
```



With a flag:

```
-------------------------------------- benchmark 'arithmetic': 1 tests ---------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     3.6224  3.6333  3.6273  0.0043  3.6268  0.0067       2;0  0.2757       5           1
--------------------------------------------------------------------------------------------------------------

-------------------------------------- benchmark 'arithmetic-compile': 1 tests --------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     19.1647  33.4298  20.5668  2.8158  19.4101  0.9302      4;11  48.6221      50           1
---------------------------------------------------------------------------------------------------------------------

---------------------------------------------- benchmark 'ini': 1 tests ----------------------------------------------
Name (time in ms)              Min       Max      Mean  StdDev    Median     IQR  Outliers     OPS  Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     627.5808  636.0194  632.0631  3.2096  631.3121  4.2531       2;0  1.5821       5           1
----------------------------------------------------------------------------------------------------------------------

------------------------------------------ benchmark 'ini-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     19.4110  33.8223  20.4391  2.4277  19.6380  0.3726       4;9  48.9258      41           1
---------------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json': 1 tests ------------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     6.6579  6.6870  6.6699  0.0134  6.6634  0.0236       1;0  0.1499       5           1
--------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     29.8766  45.6162  31.5961  3.2599  30.3657  1.1974       4;5  31.6495      32           1
---------------------------------------------------------------------------------------------------------------------
```

Always on (the previous version, without a flag):

```
-------------------------------------- benchmark 'arithmetic': 1 tests ---------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     3.7267  3.7449  3.7380  0.0068  3.7396  0.0063       2;0  0.2675       5           1
--------------------------------------------------------------------------------------------------------------

-------------------------------------- benchmark 'arithmetic-compile': 1 tests --------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     19.3040  32.5226  20.5547  2.6698  19.6383  0.3681       4;9  48.6507      50           1
---------------------------------------------------------------------------------------------------------------------

----------------------------------------------- benchmark 'ini': 1 tests ----------------------------------------------
Name (time in ms)              Min       Max      Mean  StdDev    Median      IQR  Outliers     OPS  Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     641.2466  664.4995  648.7045  9.3477  646.7174  10.5931       1;0  1.5415       5           1
-----------------------------------------------------------------------------------------------------------------------

------------------------------------------ benchmark 'ini-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     19.5595  31.3382  20.5078  1.9280  19.8379  0.8130       2;5  48.7620      50           1
---------------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json': 1 tests ------------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     6.9633  7.0019  6.9863  0.0175  6.9967  0.0290       1;0  0.1431       5           1
--------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     30.1013  42.2325  32.1045  3.1885  30.5720  2.0026       4;4  31.1483      32           1
---------------------------------------------------------------------------------------------------------------------
```

Removed asserts:

```
-------------------------------------- benchmark 'arithmetic': 1 tests ---------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     3.2740  3.3103  3.2963  0.0140  3.3010  0.0177       1;0  0.3034       5           1
--------------------------------------------------------------------------------------------------------------

-------------------------------------- benchmark 'arithmetic-compile': 1 tests --------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     18.5920  32.9804  19.8039  2.5817  18.8803  0.8268       3;8  50.4952      53           1
---------------------------------------------------------------------------------------------------------------------

---------------------------------------------- benchmark 'ini': 1 tests ----------------------------------------------
Name (time in ms)              Min       Max      Mean  StdDev    Median     IQR  Outliers     OPS  Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     565.2981  577.1831  570.2641  4.9078  570.8486  7.7442       2;0  1.7536       5           1
----------------------------------------------------------------------------------------------------------------------

------------------------------------------ benchmark 'ini-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     18.7165  29.9016  19.6981  2.0014  18.9750  0.2352      5;11  50.7662      52           1
---------------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json': 1 tests ------------------------------------------
Name (time in s)             Min     Max    Mean  StdDev  Median     IQR  Outliers     OPS  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------
test_parse_time[lark]     6.1023  6.1692  6.1446  0.0287  6.1599  0.0442       1;0  0.1627       5           1
--------------------------------------------------------------------------------------------------------------

----------------------------------------- benchmark 'json-compile': 1 tests -----------------------------------------
Name (time in ms)               Min      Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_compile_time[lark]     28.8005  41.3566  30.8045  3.1636  29.5189  1.9951       3;3  32.4628      29           1
---------------------------------------------------------------------------------------------------------------------
```